<a href="https://colab.research.google.com/github/drpedrazas/TAREAS_Aprendizaje_de_Maquina/blob/main/Tercer_Conjunto_de_Problemas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Ejercicio 2.2
#**(a)** 

**(i)** Para este caso tenemos:
$$m_{\mathcal{H}}(N) \leq \binom{N}{1} + \binom{N}{0} = N + 1$$

**(ii)** Por otro lado:
$$m_{\mathcal{H}}(N) \leq \binom{N}{2} + \binom{N}{1} + \binom{N}{0}= \frac{N^2 + N + 2}{2}$$

#**(b)**

Recuerde que si $m_{\mathcal{H}}(N) < 2^N$ existe un punto de quiebre y, de ser así, $m_{\mathcal{H}}(N)$ debe ser polinómica.

##Ejercicio 2.5
Note que 

$$|E_{out}(g) - E_{in}(g)| = 0.1$$

Luego:

$$0.1 = \sqrt{\frac{8}{N} \ln{\frac{4m_{\mathcal{H}}(2N)}{\delta}}}$$ 

Es decir:

$$\delta = e^{\frac{-(0.1)^2 N}{8}}4(201)$$

$$\delta = 709$$



#Ejercicio 2.6

**(a)** La desigualdad de Hoeffding nos da $\epsilon = 0.096$ para el conjunto de prueba y $\epsilon = 0.115$ para el conjunto de entrenamiento. Acá podemos usa la cota de Hoeffding dado que neustros conjuntos son finitos.

**(b)** con un conjunto de entrenaminto muy pequeño, se puede obtener una hipótesis que ajusta los puntos de entrenamiento de forma perfecta, pero que no nos dice mucho sobre los puntos fuera de este conjunto. Por ejemplo: cuando sólo hay un punto de entrenamiento.

##Problema 2.1

Note que a partir de $$\epsilon = \sqrt{\frac{1}{2N} \ln{\frac{2M}{\delta}}}$$
Podemos obtener $N$. Esto es:

$$N= \frac{1}{2\epsilon^2}\ln{\frac{2M}{\delta}}$$

**(a)** Para $M=1$ y $\delta = 0.05$, si $\epsilon \leq 0.05$ requerimos:

$$N \geq \frac{1}{2\cdot 0.05^2}\ln{\frac{2}{0.03}}=840$$

**(b)** Para $M=100$ y $\delta = 0.05$

$$N \geq \frac{1}{2\cdot 0.05^2}\ln{\frac{200}{0.03}}=1761$$

**(c)** Para $M=1000$ y $\delta = 0.05$

$$N \geq \frac{1}{2\cdot 0.05^2}\ln{\frac{2000}{0.03}}=2682$$

##Problema 2.12

Para este caso $d_V = 10$, $\delta = 0.05$ y $\epsilon = 0.05$.

Tenemos esta cota inferior:

$$N \geq \frac{8}{\epsilon^2} \ln \left ( 4 \frac{(2N)^{V_c} + 1}{\delta} \right )$$

Tomemos $N=500$ y realicemos un proceso iterativo simple. 



In [ ]:
import numpy as np
func = lambda x: (8/(0.05)**2)*np.log(4*((2*x)**10+1)/0.05)
n_old = 500
n_new = func(n_old)
it = 0
while abs(n_old-n_new) > 0:
  it += 1 
  n_old = n_new
  n_new = func(n_old)
print(n_old, it)

452956.8647230992 16


Luego $N=452957$

##Problema 2.16

Escogemos $D+1$ puntos distintos en $\mathbb{R}$: $x_0, \cdots , x_D$. Consirdere la matriz:

$$A=  \begin{pmatrix}1 & a_0 & \cdots & x_0^{D} \\
 \vdots &  \vdots & & \vdots \\
 1 & x_N & \cdots & x_N^{D}\end{pmatrix}$$

 Como $x_j \neq x_i$ para $i \neq j$, entonces $\det A \neq 0$, luego $A^{-1}$ existe. Considere: $y \in \{-1,1\}^{D+1}$, una dicotomía cualquiera. Existe $c$ tal que $A^{-1} y = c$; es decir: $Ac = y$, esto es:

 $$h_{s}(x_k) = signo(\sum_{i=0}^{D} c_i x_k^i)=y_k$$

 para todo $k=0, \cdots , D$. En conclusión; nuestra hipótesis $\mathcal{H}$ parte estos puntos, luego $m_{\mathcal{H}}(D+1) = 2^{D+1}$, luego $d_{VC} (\mathcal{H}) \geq D+1 $.

 Por otro lado: tome $D+2$ puntos en $\mathbb{R}$: $x_0, \cdots x_{D+1}$. Se pueden construir $D+2$ vectores $(x_k^0, x_k, \cdots, x_k^{D+1})$ : uno por cada $0 \leq k \leq D + 1 $. Estos son $D+2$ vectores en $\mathbb{R}^{D+1}$, luego es un conjunto de vectores dependiente. 
 
 Si $r_k = (x_k, \cdots , x_k^{D})$, exite $t$ tal que $r_t = \sum_{i\neq t} a_i r_i$ con no todos los $a_i$'s iguales a $0$. Sea $y_i = signo(a_i)$. Consideramos entonces $c=(c_0, \cdots, c_D)$ y vemos que:

 $$r_{t}c^{T}=(\sum_{i\neq t} a_i r_i)c^{T} = \sum_{i \neq t} \sum_{j=0}^{D} c_j a_i x_i^j$$

 Asuma que existe $c$ tal que 

 $$y_i = h_c (x_i) = signo(\sum_{j=0}^{D} c_j x_i^j)$$

 para cualquier $i$ tal que $a_i \neq 0$ y

 $$y_t = h_c(x_t)=signo(\sum_{j=0}^{D} c_j x_t^j)$$

 en este caso 

 $$signo(a_i) = y_i = signo(\sum_{j=0}^{D} c_j x_i^j)$$

 que es lo mismo que

 $$\sum_{j=0}^D c_j a_i x_i^j > 0$$
 
 Entonces $r_t c^T > 0$, luego $y_t = 1$. $y_t$ puede ser $-1$, luego hay una dicotomía que no puede ser generada por $\mathcal{H}$.